### PR TITLE
Fix boundaries in applyClampedDragUpdate

### PR DESCRIPTION
**Pull request by Hixie**

### CODE REVIEW

**Comment by Hixie** [comment]
cc @collinjackson This should fix the last customer:gold NestedScrollView bug.
Fixes https://github.com/flutter/flutter/issues/13496

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 1]
```diff
@@ -947,8 +947,10 @@ class _NestedScrollPosition extends ScrollPosition implements ScrollActivityDele
   // Returns the amount of delta that was not used.
   double applyClampedDragUpdate(double delta) {
     assert(delta != 0.0);
-    final double min = delta < 0.0 ? -double.INFINITY : minScrollExtent;
-    final double max = delta > 0.0 ? double.INFINITY : maxScrollExtent;
+    // Positive delta means going down (exposing stuff above), negative delta
+    // going up (exposing stuff below).
+    final double min = delta < 0.0 ? -double.INFINITY : math.min(minScrollExtent, pixels);
```
Comment (amirh) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: First time looking at this code, tried to snoop around and understand what's going on a bit, but I'm still lost.

Could you add a comment explaining the fix? the questions I'm currently trying to understand are: when will pixels be out of the min/max scroll extent? IIUC with the new change here in the case pixels is already out of the min/max extent we just ignore a drag that makes it go even farther away from the min/max scroll extent? why is that what we want to do?

(I'm totally aware that it is very likely that I completely misunderstood what's going here :) )

**Comment by Hixie** [comment]
@amirh Ok I've added additional comments. PTAL?

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 2]
```diff
@@ -945,10 +949,26 @@ class _NestedScrollPosition extends ScrollPosition implements ScrollActivityDele
...
 
   // Returns the amount of delta that was not used.
+  //
+  // Positive delta means going down (exposing stuff above), negative delta
+  // going up (exposing stuff below).
   double applyClampedDragUpdate(double delta) {
     assert(delta != 0.0);
-    final double min = delta < 0.0 ? -double.INFINITY : minScrollExtent;
-    final double max = delta > 0.0 ? double.INFINITY : maxScrollExtent;
+    // If we are going towards the maxScrollExtent (negative scroll offset),
```
Comment (amirh) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: The first sentence here is not adding providing more information than reading the one line of code.

What I would find useful here is a comment saying, how we get to a state where pixels is out of the min/max scroll extent, and why are we not clamping in this case.

File: packages/flutter/test/widgets/nested_scroll_view_test.dart [hunk 3]
```diff
@@ -566,4 +566,80 @@ void main() {
     expect(buildCount, expectedBuildCount);
     expect(find.byType(NestedScrollView), isNot(paints..shadow()));
   });
+
+  testWidgets('NestedScrollView and iOS bouncing', (WidgetTester tester) async {
```
Comment (amirh) [inline packages/flutter/test/widgets/nested_scroll_view_test.dart]: nit: I would add a comment explaining what is special about this sequence of event and what was the bug it's catching

**Review by amirh — APPROVED** [review]
LGTM other than 2 commenting requests

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart (hunk 2 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/nested_scroll_view.dart]: added a paragraph that explains that

File: packages/flutter/test/widgets/nested_scroll_view_test.dart (hunk 3 above)
Comment (Hixie) [reply packages/flutter/test/widgets/nested_scroll_view_test.dart]: added a comment explaining what this tests

**Comment by Hixie** [comment]
will land on green

**Pull request merged**